### PR TITLE
Connected components

### PR DESCRIPTION
**Pull request by alecjacobson**
We previously had a function that computed connected components based on triangle-triangle adjacency (defined as sharing an edge). This PR generalizes this to a routine to compute connected components from any graph (via adjacency matrix). It also adds a routine to build the facet adjacency matrix explicitly, and to do so provides a more memory efficient edge adjacency arrays.   

#### Check all that apply (change to `[x]`)
- [x] All changes meet [libigl style-guidelines](https://libigl.github.io/style-guidelines/).
- [x] Adds new .cpp file.
- [x] Adds corresponding unit test.
- [ ] This is a minor change.

### CODE REVIEW

**Comment by jdumas** [comment]
Hmm this still includes your changes from the matcap PR. Could this be separated?

**Comment by alecjacobson** [comment]
Since this probably won't be the last time I accidentally base a new PR branch off of the previous PR branch, I'll save myself some googling. What I did was:

    git checkout connected_components
    git rebase --onto master matcap connected_components
    git push --force

File: include/igl/facet_adjacency_matrix.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  Eigen::Matrix<Index,Dynamic,2> E,uE;
+  igl::unique_edge_map(F,E,uE,EMAP,uEC,uEE);
+  std::vector<Eigen::Triplet<Index> > AIJV;
+  AIJV.reserve(2*uE.rows());
+  const int nu = uE.rows();
+  for(int ue = 0;ue<nu;ue++)
+  {
+    // number of faces incident on this unique edge
+    const int mue = uEC(ue+1)-uEC(ue);
+    assert(mue <= 3 && "DEBUGGING, DELETE THIS ASSERTION");
```
Comment (jdumas) [inline include/igl/facet_adjacency_matrix.cpp:29]: ?

File: include/igl/facet_adjacency_matrix.cpp [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+template <typename DerivedF, typename Atype>
+IGL_INLINE void igl::facet_adjacency_matrix(
+  const Eigen::MatrixBase<DerivedF> & F,
+  Eigen::SparseMatrix<Atype> & A)
+{
+  using namespace Eigen;
+  typedef typename DerivedF::Scalar Index;
+  const int m = F.rows();
+  Eigen::Matrix<Index,Dynamic,1> EMAP,uEE,uEC;
```
Comment (jdumas) [inline include/igl/facet_adjacency_matrix.cpp:19]: When I see this, I really wish we'd use more explicit variable names... (It's C++, not MATLibberish :p)

File: include/igl/matlab/prepare_lhs.h [hunk 3]
```diff
@@ -40,6 +40,11 @@ namespace igl
     IGL_INLINE void prepare_lhs_double(
       const Eigen::SparseMatrix<Vtype> & V,
       mxArray *plhs[]);
+    // Vector of matrices -> cell array of matrices
+    template <typename Vtype>
+    IGL_INLINE void prepare_lhs_double(
```
Comment (jdumas) [inline include/igl/matlab/prepare_lhs.h:45]: No implementation for this guy?

File: include/igl/unique_edge_map.cpp [hunk 4]
```diff
@@ -33,19 +61,54 @@ IGL_INLINE void igl::unique_edge_map(
...
+  assert(uEK.rows()+1 == uEC.rows());
+  // running inner offset in uEE
+  VectorXI uEO = VectorXI::Zero(uE.rows(),1);
+  // flat array of faces incide on each uE
+  uEE.resize(EMAP.rows(),1);
+  for(int e = 0;e<EMAP.rows();e++)
   {
-    uE2E[EMAP(e)].push_back(e);
+    const int ue = EMAP(e);
+    const int i = uEC(ue)+ uEO(ue);
```
Comment (jdumas) [inline include/igl/unique_edge_map.cpp:100]: Type should be `DeriveduEC::Scalar` ?

File: include/igl/facet_adjacency_matrix.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  Eigen::Matrix<Index,Dynamic,2> E,uE;
+  igl::unique_edge_map(F,E,uE,EMAP,uEC,uEE);
+  std::vector<Eigen::Triplet<Index> > AIJV;
+  AIJV.reserve(2*uE.rows());
+  const int nu = uE.rows();
```
Comment (jdumas) [inline include/igl/facet_adjacency_matrix.cpp:24]: All those sizes should be `Eigen::Index` (= `long`) rather than `int`.

**Comment by alecjacobson** [comment]
Sure, that's fine. I don't mind changing those types (also don't mind if you do minor edits directly).

**Comment by jdumas** [comment]
Sure thing. I've pushed more int stuff. Should be good to go now.

**Comment by alecjacobson** [comment]
btw @qnzhou , we're using the old, slower `unique_edge_map` in `mesh_boolean`. We could replace accordingly with this faster version.

**Pull request merged**